internal/migrations/changesets.go

```diff
@@ -0,0 +1,319 @@
+	ConfirmationThreshold: big.NewRat(2, 3),
+	ExpirationPeriod:      24 * 60 * 10, // 1 day in blocks
+	ResolveFunc: func(ctx context.Context, app *common.App, resolution *resolutions.Resolution, block *common.BlockContext) error {
+
+		// Extract the migration config from the app, executed only once
+		if migrationCfg == nil {
+			migrationCfg = &MigrationConfig{}
+			cfg, ok := app.Service.ExtensionConfigs[ListenerName]
+			if !ok {
+				return errors.New("no migration config provided")
+			}
+
+			err := migrationCfg.ExtractConfig(cfg)
+			if err != nil {
+				return err
+			}
+		}
+
+		var migration ChangesetMigration
+		err := migration.UnmarshalBinary(resolution.Body)
+		if err != nil {
+			return err
+		}
+
+		// insert the changeset into the database
+		app.Service.Logger.Debug("insert changeset chunk", log.Int("height", migration.Height.Int64()), log.Int("chunkIndex", migration.ChunkIdx.Int64()))
+		tx, err := app.DB.BeginTx(ctx)
+		if err != nil {
+			return err
+		}
+		defer tx.Rollback(ctx)
+
+		// insert the changeset into the database
+		err = migration.insertChangeset(ctx, tx)
+		if err != nil {
+			return err
+		}
+
+		// Commit the changeset chunk to the database, so that it is persisted
+		// even if the resolution fails later
+		if err = tx.Commit(ctx); err != nil {
+			return err
+		}
+
+		tx, err = app.DB.BeginTx(ctx)
+		if err != nil {
+			return err
+		}
+		defer tx.Rollback(ctx)
+
+		var currentHeight int64
+		// get the last changeset height applied
+		lastChangeset, err := getLastChangeset(ctx, tx)
+		if err != nil {
+			return tx.Commit(ctx)
+		}
+
+		if lastChangeset == migrationCfg.EndHeight {
+			return nil
+		} else if lastChangeset == -1 {
+			currentHeight = migrationCfg.StartHeight
+		} else {
+			currentHeight = lastChangeset + 1
+		}
+
+		// Apply the changesets in order of block height
+		for {
+			// If the current height is greater than the migration end height, break
+			if currentHeight >= migrationCfg.EndHeight {
+				app.Service.Logger.Info("changeset migration completed", log.Int("height", currentHeight))
+				break
+			}
+
+			// Check if all chunks have been received for the current height, if not, break
+			totalChunks, chunksReceived, err := getChangesetMetadata(ctx, tx, currentHeight)
+			if err != nil {
+				break
+			}
+
+			if totalChunks != chunksReceived {
+				// app.Service.Logger.Info("waiting for all chunks to be received", log.Int("height", currentHeight))
+				break
+			}
+
+			// Apply the changeset
+			if err = applyChangeset(ctx, tx, currentHeight, totalChunks); err != nil {
+				return err
+			}
+
+			app.Service.Logger.Info("Applied changesets", log.Int("height", currentHeight))
+
+			// Delete the changeset after it has been applied
+			if err = deleteChangesets(ctx, tx, currentHeight); err != nil {
+				return err
+			}
+
+			// Increment the last changeset
+			if err = setLastChangeset(ctx, tx, currentHeight); err != nil {
```

Comment:
I don't understand why "last applied changeset" is a thing to track at all.  This loop and the transaction it uses is supposed to make the application of the changesets all-or-none, right?  Any early return rolls it all back.

Reply:
This loop tries to apply all the block heights for which all the chunks have been received. That's why we need to track the last applied block height